raftstore: always update read delegate's region

overvenus opened the pull request:
Signed-off-by: Neil Shen <overvenus@gmail.com>

## What have you changed? (mandatory)

Always update read delegate's region to avoid stale region info in the localreader. So read delegates always have up to data region when their peer become leader.

## What are the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

Unit tests.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#2639

Connor1996 commented on src/raftstore/store/peer.rs:
```diff
@@ -539,7 +539,8 @@ impl Peer {
         self.mut_store().set_region(region.clone());
         let progress = ReadProgress::region(region);
-        self.maybe_update_read_progress(progress);
+        // Always update read delegate's region.
```
maybe you should add a comment for reason why we always update it.

Connor1996 commented on src/raftstore/store/worker/read.rs:
```diff
@@ -358,6 +361,7 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
             Ok(RequestPolicy::ReadLocal) => Ok(Some(delegate)),
             // It can not handle other policies.
             Ok(_) => Ok(None),
+
```
please remove blank line

Connor1996 commented on src/raftstore/store/worker/read.rs:
```diff
@@ -103,16 +103,17 @@ impl ReadDelegate {
-                        cmd_resp::bind_term(&mut resp.response, term);
-                        return Some(resp);
-                    }
+                    let mut resp = executor.execute(req, &self.region);
```
if `check_region_epoch()` failed in `execute()`, now `handle_read()` will return `Some(resp with snap is None)` and before will return None, is that alright?

overvenus replied:
Yes, it is okay, because we disable epoch check in localreader.

BusyJay replied:
I think this will become a pitfall in the future.

overvenus replied:
Could you make an example?

BusyJay replied:
It's not obvious for reader. If a reviewer misunderstood it, people could make the same mistake too when refactoring the code.

BusyJay commented:
Please describe the bug you are trying to fix in the description.

BusyJay commented:
What if the request arrives before region is updated?

overvenus commented:
It may be handled or rejected by lease, it depends to what kind of changes made to the region.

There are two categories of commands that changes region:
 1. conf changes, they have no effect to read delegates as long as theirs peers are leaders.
 2. split / merge
   - split: the request maybe handled, because peers of new region that in other store do not start election before theirs election timeout which is longer than the max leader lease
   - merge: we suspect lease right after prepare merge is committed, so the request is rejected.

BusyJay commented on src/raftstore/store/worker/read.rs:
```diff
@@ -325,10 +327,10 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
             }
             None => {
                 self.metrics.borrow_mut().rejected_by_no_region += 1;
+                debug!("rejectd by no region {}", region_id);
```
s/rejectd/rejected/

BusyJay commented on src/raftstore/store/worker/read.rs:
```diff
@@ -350,6 +352,7 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
         if util::check_region_epoch(req, &delegate.region, false).is_err() {
             self.metrics.borrow_mut().rejected_by_epoch += 1;
             // Stale epoch, redirect it to raftstore to get the latest region.
+            debug!("{} rejectd by stale epoch", delegate.tag);
```
Ditto.

overvenus commented:
PTAL, thanks!

hicqu commented:
LGTM.

overvenus commented:
/run-integration-tests
